### PR TITLE
21535 - Eft refund for eft_credits

### PR DESCRIPTION
**Pull request by Jxio**
*Issue #:*
https://github.com/bcgov/entity/issues/21535

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

**Review by seeker25 — APPROVED** [review]

File: pay-api/src/pay_api/services/eft_service.py [hunk 1]
```diff
@@ -121,14 +123,40 @@ def create_shortname_refund(cls, request: Dict[str, str], **kwargs) -> Dict[str,
...
         html_body = cls._render_email_body(shortname, amount, comment, shortname_id)
 
         send_email(recipients, subject, html_body, **kwargs)
         refund.save()
 
+    @classmethod
+    def _refund_eft_credits(cls, shortname_id: str, amount: str):
+        """Refund the amount to eft_credits table based on short_name_id."""
+        refund_amount = Decimal(amount)
+        eft_credits = EFTCreditModel.find_by_short_name_id(int(shortname_id))
```
Comment (ochiu) [inline pay-api/src/pay_api/services/eft_service.py:139]: EFTShortnames service has a couple of functions that might be useful
`def get_eft_credit_balance(short_name_id: int) -> Decimal:`
`def get_eft_credits(short_name_id: int) -> List[EFTCreditModel]:`

I would add validation to check if the refund amount is greater than the credit balance currently and throw business exception

File: pay-api/src/pay_api/services/eft_service.py [hunk 2]
```diff
@@ -121,14 +123,40 @@ def create_shortname_refund(cls, request: Dict[str, str], **kwargs) -> Dict[str,
...
+                credit.remaining_amount -= refund_amount
+                refund_amount = Decimal(0)
+            else:
+                refund_amount -= credit_amount
+                credit.remaining_amount = Decimal(0)
+
+            credit.save()
+
+        if refund_amount > 0:
+            current_app.logger.error(f'Shortname {shortname_id} Refund amount exceed eft_credits remaining amount.')
```
Comment (ochiu) [inline pay-api/src/pay_api/services/eft_service.py:158]: At this point you may have already done a save() even though there is not enough credits. We should probably roll back if conditions aren't met? My suggestion for the validation above should prevent this.

Should add a test for the business exception

File: pay-api/src/pay_api/services/eft_service.py [hunk 3]
```diff
@@ -121,14 +126,39 @@ def create_shortname_refund(cls, request: Dict[str, str], **kwargs) -> Dict[str,
...
+    @classmethod
+    def _refund_eft_credits(cls, shortname_id: int, amount: str):
+        """Refund the amount to eft_credits table based on short_name_id."""
+        refund_amount = Decimal(amount)
+        eft_credits: List[EFTCreditModel] = EFTShortnames.get_eft_credits(shortname_id)
+        eft_credit_balance = EFTShortnames.get_eft_credit_balance(shortname_id)
+
+        if refund_amount > eft_credit_balance:
+            current_app.logger.error(f'Shortname {shortname_id} Refund amount exceed eft_credits remaining amount.')
+            raise BusinessException(Error.INVALID_TRANSACTION)
```
Comment (ochiu) [inline pay-api/src/pay_api/services/eft_service.py:147]: I personally would add a new Error for this case, but I will leave it up to you.

Do we need the logging if we are already raising an exception?

**Review by ochiu — APPROVED** [review]
Looks good, thanks

File: pay-api/src/pay_api/services/eft_service.py [hunk 4]
```diff
@@ -121,14 +126,38 @@ def create_shortname_refund(cls, request: Dict[str, str], **kwargs) -> Dict[str,
...
         html_body = cls._render_email_body(shortname, amount, comment, shortname_id)
 
         send_email(recipients, subject, html_body, **kwargs)
         refund.save()
 
+    @classmethod
+    def _refund_eft_credits(cls, shortname_id: int, amount: str):
+        """Refund the amount to eft_credits table based on short_name_id."""
+        refund_amount = Decimal(amount)
+        eft_credits: List[EFTCreditModel] = EFTShortnames.get_eft_credits(shortname_id)
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:142]: Low priority:

This typehint isn't needed, function already has the hint defined

**Pull request merged**